File: samples/python/streaming-sample/streaming_sample.py
import asyncio
import sys
from aeron_cache.client import AeronCacheClient
from aeron_cache.embedded_cache import EmbeddedAeronCache
from aeron_cache.embedded_counter_cache import EmbeddedCounterCache

async def main():
    base_url = "http://localhost:7070"
    ws_url = "ws://localhost:7071"
    if len(sys.argv) > 1:
        base_url = sys.argv[1]
        ws_url = sys.argv[2]

    print(f"Starting Streaming Sample against {base_url}, ws={ws_url}")

    client = AeronCacheClient(base_url, ws_url)

    # Ensure cache exists
    try:
        await client.create_cache_async("streaming-sample-cache")
    except:
        pass

    cache = EmbeddedAeronCache(client, "streaming-sample-cache")
    print("Connected. Waiting for updates on 'streaming-sample-cache'...")

    async def on_changes(data):
        print(f"[Python] Observed change: {data}")

    async def poller():
        last_val = None
        while True:
            val = cache.get_local("streaming-key")
            if val and val != last_val:
                print(f"[Python-Poller] 'streaming-key' updated: {val}")
                last_val = val
            await asyncio.sleep(1)

    asyncio.create_task(poller())

    # --- Counter streaming (with hydration) ---
    try:
        await client.create_counter_cache_async("streaming-counter-cache")
    except Exception:
        pass

    counters = EmbeddedCounterCache(client, "streaming-counter-cache")

    async def on_counter_change(event):
        print(f"[Python] Counter update: {event.eventType} {event.itemKey} -> {event.itemValue}")

    async def counter_ticker():
        while True:
            try:
                await counters.increment_async("tick", 1)
            except Exception:
                pass
            await asyncio.sleep(2)

    asyncio.create_task(counters.subscribe(on_counter_change, hydrate=True))
    asyncio.create_task(counter_ticker())

    # This will block forever receiving updates, with initial hydration
    print("Subscribing with hydration...")
    await cache.subscribe(on_changes, hydrate=True)

if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        pass

File: samples/java/streaming-sample/src/main/java/com/aeron/cache/sample/StreamingSample.java
package com.aeron.cache.sample;


import com.bhf.aeroncache.client.AeronCacheClient;
import com.bhf.aeroncache.client.AeronCacheSubscriber;
import com.bhf.aeroncache.client.CounterCacheSubscriber;
import com.bhf.aeroncache.client.EmbeddedAeronCache;
import com.bhf.aeroncache.client.EmbeddedCounterCache;
import com.bhf.aeroncache.models.CacheUpdateEvent;
import com.bhf.aeroncache.models.CounterUpdateEvent;
import java.net.http.WebSocket;

public class StreamingSample {
    public static void main(String[] args) {
        String baseUrl = "http://localhost:7070";

        System.out.println("Starting Streaming Sample against " + baseUrl);

        var wsUrl = "ws://localhost:7071";
        AeronCacheClient client = new AeronCacheClient(baseUrl, wsUrl);
        try {
            var response = client.createCache("streaming-sample-cache");
            System.out.println("Created cache: " + response.getCacheId());
        } catch (Exception e) {
            // Probably already exists
        }
        EmbeddedAeronCache cache = new EmbeddedAeronCache(client, "streaming-sample-cache");

        // Subscribe explicitly using AeronCacheSubscriber abstraction with hydration
        cache.subscribe(new AeronCacheSubscriber() {
            @Override
            public void onOpen(WebSocket webSocket) {
                super.onOpen(webSocket);
                System.out.println("[Java] WebSocket subscription opened (with hydration)");
            }

            @Override
            public void onAfterUpdate(CacheUpdateEvent event) {
                System.out.println("[Java] Received update: " + event.getEventType() + " for " + event.getItemKey());
            }
        }, true);

        // --- Counter streaming (with hydration) ---
        try {
            var response = client.createCounterCache("streaming-counter-cache");
            System.out.println("Created counter cache: " + response.getCacheId());
        } catch (Exception e) {
            // Probably already exists
        }
        EmbeddedCounterCache counters = new EmbeddedCounterCache(client, "streaming-counter-cache");
        counters.subscribe(new CounterCacheSubscriber() {
            @Override
            public void onAfterUpdate(CounterUpdateEvent event) {
                System.out.println("[Java] Counter update: " + event.getEventType()
                        + " for " + event.getItemKey() + " -> " + event.getItemValue());
            }
        }, true);

        // Periodically increment a counter to generate streaming updates.
        Thread ticker = new Thread(() -> {
            while (true) {
                try {
                    counters.increment("tick", 1);
                    Thread.sleep(2000);
                } catch (Exception e) {
                    return;
                }
            }
        });
        ticker.setDaemon(true);
        ticker.start();

        System.out.println("Listening for updates on 'streaming-sample-cache'.");

        String lastValue = "";
        while (true) {
            // Poll using getLocal instead of get
            String current = cache.getLocal("streaming-key");
            
            if (current != null && !current.isEmpty() && !current.equals(lastValue)) {
                System.out.println("[Java-Poller] Observed change in 'streaming-key': " + current);
                lastValue = current;
            }

            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }
    }
}
